### PR TITLE
Add group policy for sync service URL

### PR DESCRIPTION
**Pull request by bsclifton**
Fixes https://github.com/brave/brave-browser/issues/20431

Security review https://github.com/brave/reviews/issues/1744

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used GitHub [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

#### Using regedit.exe
1. Open `regedit.exe`
2. Navigate to `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\`
3. Create the key (folders) `BraveSoftware\Brave\` if they don't already exist
4. Create a new string value (`REG_SZ`) with the name `BraveSyncUrl` and the value `https://sync-v2.brave.com/v2`. 
5. Load Brave; verify it shows under brave://policy/
  
#### Adding using a `.reg` file (also on Windows)
1. Create a new empty file called `sync-policy.reg` 
2. Open it in Notepad and put this for the content:
```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Policies\BraveSoftware\Brave]
"BraveSyncUrl"="https://sync-v2.brave.com/v2"
```
3. Save the file and close it
4. Double click the `sync-policy.reg` file to run it (you should get prompted)
5. Load Brave; verify it shows under brave://policy/

### macOS
You can set via the command line. 

- For Brave, you'd use a bundle ID of `com.brave.Browser`, `com.brave.Browser.beta`, or `com.brave.Browser.nightly` (depending on your channel. 
- The key name will be `BraveSyncUrl`
- The value would be the sync server. For example, `https://sync-v2.brave.com/v2`

Steps for Release channel (change bundle ID for other channels)

1. Run the following from command line:
    ```
    defaults write com.brave.Browser BraveSyncUrl -string https://sync-v2.brave.com/v2
    ```

2. Load Brave; verify it shows under brave://policy/

### Linux
You can set the options via `/etc/brave/policies/managed`

Basically, you will make a file with JSON matching the keys/values you'd like to set. Full overview here:
https://source.chromium.org/chromium/chromium/src/+/main:docs/website/site/administrators/linux-quick-start/index.md

### Example of how to set the sync URL
1. create a file 
    ```bash
    vim /etc/brave/policies/managed/sync_url_policy.json
    ```
2. paste in:
    ```json
    {
        "BraveSyncUrl": "https://sync-v2.brave.com/v2"
    }
    ```

3. Save the file and exit (escape; `:wq`, enter)
4. Load Brave; verify it shows under brave://policy/
5. Delete the file after testing is complete

### CODE REVIEW

File: chromium_src/components/sync/service/sync_service_impl.cc [hunk 1]
```diff
@@ -3,13 +3,50 @@
...
+                            PrefService* prefs) {
+  // Allow group policy to override sync service URL.
+  // This has a higher priority than the --sync-url command-line param.
+  // https://github.com/brave/brave-browser/issues/20431
+  if (prefs) {
+    if (prefs->IsManagedPreference(brave_sync::kManagedBraveSyncUrl)) {
+      std::string value(prefs->GetString(brave_sync::kManagedBraveSyncUrl));
+      if (!value.empty()) {
+        GURL custom_sync_url(value);
+        if (custom_sync_url.is_valid()) {
```
Comment (fmarier) [inline chromium_src/components/sync/service/sync_service_impl.cc:25]: Let's make sure it's an HTTPS URL too.

File: chromium_src/components/policy/tools/generate_policy_source.py [hunk 2]
```diff
@@ -166,6 +166,26 @@ def AddBravePolicies(template_file_contents):
             'desc': ('''This policy allows an admin to specify that Brave '''
                      '''AI Chat feature will be enabled.'''),
         },
+        {
+            'name': 'BraveSyncUrl',
+            'type': 'main',
+            'schema': {
+                'type': 'string'
+            },
+            'supported_on': ['chrome.*:128-'],
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/components/policy/tools/generate_policy_source.py]: Yes! Thanks 😄 Fixed

File: chromium_src/components/sync/service/sync_service_impl.cc (hunk 1 above)
Comment (bsclifton) [reply chromium_src/components/sync/service/sync_service_impl.cc]: Fixed!

File: components/brave_sync/brave_sync_prefs.h [hunk 3]
```diff
@@ -23,6 +23,8 @@ class Time;
 
 namespace brave_sync {
 
+inline constexpr char kManagedBraveSyncUrl[] = "brave_sync.sync_url";
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply components/brave_sync/brave_sync_prefs.h]: Fixed 👍

**Comment by bsclifton** [comment]
OK pushed fixes - however, I am getting a `presubmit` error that I am not sure how to solve:
```
** Presubmit ERRORS: 1 **
You added one or more #includes that violate checkdeps rules.
  brave\chromium_src\components\sync\service\sync_service_impl.cc
      Illegal include: "brave/components/brave_sync/brave_sync_prefs.h"
    Because of no rule applying.
```

It looks like this should be fine...
- we patch upstream to add sources/deps here: https://github.com/brave/brave-core/blob/91d59fae31e22a9b043bc5b8c8c019cc371a501b/patches/components-sync-service-BUILD.gn.patch#L9
- `//brave/components/brave_sync:prefs` is called out in https://github.com/brave/brave-core/blob/91d59fae31e22a9b043bc5b8c8c019cc371a501b/components/sync/service/sources.gni#L19

**Review by goodov — APPROVED** [review]

File: chromium_src/components/sync/service/sync_service_impl.cc [hunk 4]
```diff
@@ -3,13 +3,53 @@
...
+namespace syncer {
+
+GURL BraveGetSyncServiceURL(const base::CommandLine& command_line,
+                            version_info::Channel channel,
+                            PrefService* prefs) {
+  // Allow group policy to override sync service URL.
+  // This has a higher priority than the --sync-url command-line param.
+  // https://github.com/brave/brave-browser/issues/20431
+  if (prefs) {
+    if (prefs->IsManagedPreference(brave_sync::kCustomSyncServiceUrl)) {
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/components/sync/service/sync_service_impl.cc]: Fixed! 😄

**Review by AlexeyBarabash — APPROVED** [review]
looks good

File: chromium_src/components/sync/service/sync_service_impl.cc [hunk 5]
```diff
@@ -3,13 +3,51 @@
...
+  // See `GetSyncServiceURL` in `components/sync/base/sync_util.cc`
+  return GetSyncServiceURL(command_line, channel);
+}
+
+}  // namespace syncer
 
 #define SyncAuthManager BraveSyncAuthManager
 #define SyncStoppedReporter BraveSyncStoppedReporter
+#define GetSyncServiceURL(...) \
+  BraveGetSyncServiceURL(__VA_ARGS__, sync_client_->GetPrefService())
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/components/sync/service/sync_service_impl.cc:47]: Thanks again for your help here! 😄

**Comment by bsclifton** [comment]
Updated to add test plan for Windows / macOS / Linux.

Trying to find an Android group policy example....

**Review by fmarier — APPROVED** [review]
